### PR TITLE
On Fn arg mismatch for a fn path, suggest a closure

### PR DESCRIPTION
**Pull request by estebank**
When encountering a fn call that has a path to another fn being passed in, where an `Fn` impl is expected, and the arguments differ, suggest wrapping the argument with a closure with the appropriate arguments.

The last `help` is new:

```
error[E0631]: type mismatch in function arguments
  --> $DIR/E0631.rs:9:9
   |
LL |     fn f(_: u64) {}
   |     ------------ found signature defined here
...
LL |     foo(f);
   |     --- ^ expected due to this
   |     |
   |     required by a bound introduced by this call
   |
   = note: expected function signature `fn(usize) -> _`
              found function signature `fn(u64) -> _`
note: required by a bound in `foo`
  --> $DIR/E0631.rs:3:11
   |
LL | fn foo<F: Fn(usize)>(_: F) {}
   |           ^^^^^^^^^ required by this bound in `foo`
help: consider wrapping the function in a closure
   |
LL |     foo(|arg0: usize| f(/* u64 */));
   |         +++++++++++++  +++++++++++
```

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -2015,6 +2025,152 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+                && expected_inputs.len() == found_inputs.len()
+                && let hir::QPath::Resolved(_, path) = path
+                && let hir::def::Res::Def(_, fn_def_id) = path.res
+                && let Some(node) = self.tcx.hir().get_if_local(fn_def_id)
+                && let Some(body_id) = node.body_id()
+            {
+                let closure = self
+                    .tcx
+                    .hir()
+                    .body_param_names(body_id)
```
Comment (petrochenkov) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:2127]: `query fn_arg_names` looks like a better version of this function, it works in more cases including functions from other crates.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 2]
```diff
@@ -2015,6 +2025,152 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+            }
+        };
+        let (closure_names, call_names): (Vec<_>, Vec<_>) =
+            if both_tys.clone().all(|(expected, found)| {
+                let (expected_ty, _) = get_deref_type_and_refs(expected);
+                let (found_ty, _) = get_deref_type_and_refs(found);
+                self.can_eq(param_env, found_ty, expected_ty)
+            }) && !expected_inputs.is_empty()
+                && expected_inputs.len() == found_inputs.len()
+                && let hir::QPath::Resolved(_, path) = path
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs]: Only imported `Res`, as `hir::Enum::Variant` is quite common throughout the codebase so I wasn't sure if you meant also bringing `QPath` and the like into scope.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -2047,39 +2047,34 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
                     name
                 } else if found_refs.len() > expected_refs.len() {
-                    if found_refs[..found_refs.len() - expected_refs.len()]
-                        .iter()
-                        .zip(expected_refs.iter())
-                        .any(|(e, f)| e != f)
-                    {
-                        // The refs have different mutability.
+                    let refs = &found_refs[..found_refs.len() - expected_refs.len()];
+                    if found_refs[..expected_refs.len()].iter().eq(expected_refs.iter()) {
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:2065]: By changing it from `zip` to `eq`, now I can't rely on stopping at the shortest one, so I need this second sub-slicing of `found_refs`. Deduplicated the sub-slicing withing the two if branches though.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 4]
```diff
@@ -2015,6 +2025,152 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+        };
+        let (closure_names, call_names): (Vec<_>, Vec<_>) =
+            if both_tys.clone().all(|(expected, found)| {
+                let (expected_ty, _) = get_deref_type_and_refs(expected);
+                let (found_ty, _) = get_deref_type_and_refs(found);
+                self.can_eq(param_env, found_ty, expected_ty)
+            }) && !expected_inputs.is_empty()
+                && expected_inputs.len() == found_inputs.len()
+                && let hir::QPath::Resolved(_, path) = path
+                && let hir::def::Res::Def(_, fn_def_id) = path.res
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs]: Only imported Res, as hir::Enum::Variant is quite common throughout the codebase so I wasn't sure if you meant also bringing QPath and the like into scope.

[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs]: Ok.

**Pull request merged**